modules/navier_stokes/src/kernels/INSMomentumSUPG.C

```diff
@@ -0,0 +1,312 @@
+  if (_mu[_qp] < std::numeric_limits<double>::epsilon())
+  {
+    alpha = 1;
+    d_alpha_d_vel_component = 0;
+  }
+  // Pure diffusion
+  else if (U.norm() < std::numeric_limits<double>::epsilon())
+  {
+    alpha = 0;
+    return 0;
+  }
+  else
+  {
+    Real grid_Pe = _current_elem->hmax() * U.norm() / (2. * _mu[_qp]);
+    Real d_grid_Pe_d_vel_component =
+        _current_elem->hmax() * _u[_qp] * _phi[_j][_qp] / (2. * _mu[_qp] * U.norm());
+    alpha = 1. / std::tanh(grid_Pe) - 1. / grid_Pe;
+    d_alpha_d_vel_component =
+        ((1. - 1. / (std::tanh(grid_Pe) * std::tanh(grid_Pe))) + 1. / (grid_Pe * grid_Pe)) *
+        d_grid_Pe_d_vel_component;
+  }
+  Real PG_test = alpha * _current_elem->hmax() / 2. * U / U.norm() * _grad_test[_i][_qp];
+  RealVectorValue d_U_d_vel_component(0, 0, 0);
+  d_U_d_vel_component(_component) = _phi[_j][_qp];
+  Real d_PG_test_d_vel_component =
+      _current_elem->hmax() / 2. * _grad_test[_i][_qp] *
+      (d_alpha_d_vel_component * U / U.norm() +
+       alpha * (d_U_d_vel_component / U.norm() -
+                U / (U.norm() * U.norm() * U.norm()) * _u[_qp] * _phi[_j][_qp]));
+
+  Real convective_part =
+      _rho[_qp] * ((U * _grad_phi[_j][_qp]) + _phi[_j][_qp] * _grad_u[_qp](_component)) * PG_test;
+  convective_part += _rho[_qp] * U * _grad_u[_qp] * d_PG_test_d_vel_component;
+
+  Real pressure_part = _grad_p[_qp](_component) * d_PG_test_d_vel_component;
+
+  Real body_force_part = -_rho[_qp] * _gravity(_component) * d_PG_test_d_vel_component;
+
+  return convective_part + pressure_part + body_force_part;
+}
+
+Real
+INSMomentumSUPG::computeQpOffDiagJacobian(unsigned jvar)
+{
+  if (jvar == _u_vel_var_number)
+  {
+    RealVectorValue U(_u_vel[_qp], _v_vel[_qp], _w_vel[_qp]);
+    Real alpha;
+    Real d_alpha_d_vel_component;
+    if (_mu[_qp] < std::numeric_limits<double>::epsilon())
+    {
+      alpha = 1;
+      d_alpha_d_vel_component = 0;
+    }
+    else if (U.norm() < std::numeric_limits<double>::epsilon())
+    {
+      alpha = 0;
+      return 0;
+    }
+    else
+    {
+      Real grid_Pe = _current_elem->hmax() * U.norm() / (2. * _mu[_qp]);
+      Real d_grid_Pe_d_vel_component =
+          _current_elem->hmax() * _u_vel[_qp] * _phi[_j][_qp] / (2. * _mu[_qp] * U.norm());
+      alpha = 1. / std::tanh(grid_Pe) - 1. / grid_Pe;
+      d_alpha_d_vel_component =
+          ((1. - 1. / (std::tanh(grid_Pe) * std::tanh(grid_Pe))) + 1. / (grid_Pe * grid_Pe)) *
+          d_grid_Pe_d_vel_component;
+    }
+    Real PG_test = alpha * _current_elem->hmax() / 2. * U / U.norm() * _grad_test[_i][_qp];
+    RealVectorValue d_U_d_vel_component(0, 0, 0);
+    d_U_d_vel_component(0) = _phi[_j][_qp];
+    Real d_PG_test_d_vel_component =
+        _current_elem->hmax() / 2. * _grad_test[_i][_qp] *
+        (d_alpha_d_vel_component * U / U.norm() +
+         alpha * (d_U_d_vel_component / U.norm() -
+                  U / (U.norm() * U.norm() * U.norm()) * _u_vel[_qp] * _phi[_j][_qp]));
```

Comment:
The only difference between this line and the other components is `_u_vel` vs. `_v_vel` vs. `_w_vel`, right? Would it be possible to reduce some of the code duplication by mapping `jvar` into a `compnent` and using `U(component)` instead?